Add wrapper that applies minmod TCI to several tensors

fmahebert opened the pull request:
## Proposed changes

Adds a wrapper checks whether the minmod TCI triggers for any ~component of the Variables~ edit: of several input tensors, which can be used to trigger WENO or other more sophisticated limiters.

Depends on #1433  

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
MinmodTci: Move small functions to unnamed namespace

fmahebert committed:
Factor some Minmod functions into helpers file

fmahebert committed:
Add MinmodTci for several tensors

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/MinmodTci.hpp:
```diff
@@ -46,5 +70,142 @@ bool troubled_cell_indicator(
+// intermediate results are passed "up" out of this function.
+//
+// REQUIRES that PackagedData have a `means` and `element_size` for each
+// neighbor, as in SlopeLimiters::Minmod.
```
I should document what types these are expected to have, in addition to the names given now.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/MinmodTci.hpp:
```diff
@@ -46,5 +70,142 @@ bool troubled_cell_indicator(
+    }
+  }
+  return result;
+}
```
These three helper functions (`allocate_buffers`, `compute_effective_neighbor_sizes`, `compute_effective_neighbor_means`) don't need to be in the MinmodTci file. I will move them into a MinmodHelpers file, for a cleaner organization.

fmahebert commented:
I've force-pushed some adjustments as new commits, because no one had looked at this yet. ~This PR is ready to review.~

Edit: This will probably need still more adjustments, marking as "in progress" for now.

fmahebert commented:
Updated with new interface (originally: check a variables; now: check tensors).

Now actually ready to review :slightly_smiling_face:

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_MinmodTci.cpp:
```diff
@@ -729,11 +735,84 @@ void test_minmod_tci_3d() noexcept {
+      element, mesh, element_size);
+  CHECK(activation);
+
+  // Case where the vector triggers limiting
```
It would be good to test that any of the 3 components of the vector cause the limiter to trigger

nilsdeppe reviewed (changes requested):
I just have one request of adding another couple tests

fmahebert committed:
Avoid copying an object when preparing Minmod calls

fmahebert commented:
- rebased
- made some small improvements to the includes (was led astray by IWYU) that I squashed
- added new commit with tiny improvement, removing an unneeded copy of some data
- added the requested test as a fixup

nilsdeppe commented:
Could you please rebase this now that the dependencies have been merged?

fmahebert commented:
Rebased

nilsdeppe reviewed (commented):
:+1: go ahead and squash :)

fmahebert commented:
Rebased and squashed
